File: js/src/capabilities.ts
/** Capability negotiation helper for GSP. */

/**
 * Per-member capability advertisement plus intersection / union queries
 * for negotiating optional features (codecs, extensions, version flags).
 */
export class CapabilitiesNegotiator {
    private readonly advertised = new Map<number, Set<string>>();

    /** Record an advertisement (replaces any prior one). */
    advertise(memberId: number, capabilities: Iterable<string>): void {
        this.advertised.set(memberId, new Set(capabilities));
    }

    /** Remove a member's advertisement. */
    forget(memberId: number): void { this.advertised.delete(memberId); }

    /** Current advertisement for `memberId` (a copy). */
    capabilitiesOf(memberId: number): Set<string> | null {
        const s = this.advertised.get(memberId);
        return s ? new Set(s) : null;
    }

    /** `true` iff every advertised member supports `capability`. */
    groupSupports(capability: string): boolean {
        if (this.advertised.size === 0) return false;
        for (const s of this.advertised.values()) if (!s.has(capability)) return false;
        return true;
    }

    /** Intersection — capabilities every member advertises (safe-to-use set). */
    intersection(): Set<string> {
        const iter = this.advertised.values();
        const first = iter.next();
        if (first.done) return new Set();
        const acc = new Set(first.value);
        let n = iter.next();
        while (!n.done) {
            for (const c of acc) if (!n.value.has(c)) acc.delete(c);
            n = iter.next();
        }
        return acc;
    }

    /** Union — every capability advertised by any member. */
    union(): Set<string> {
        const acc = new Set<string>();
        for (const s of this.advertised.values()) for (const c of s) acc.add(c);
        return acc;
    }

    /** Members that did not advertise `capability`. */
    missing(capability: string): number[] {
        const out: number[] = [];
        for (const [m, s] of this.advertised) if (!s.has(capability)) out.push(m);
        return out;
    }

    get size(): number { return this.advertised.size; }
}

File: python/gbp_stack/capabilities.py
"""Capability negotiation helper for GSP."""

from __future__ import annotations

from typing import Dict, Iterable, List, Optional, Set


class CapabilitiesNegotiator:
    """Per-member capability advertisement plus intersection / union queries."""

    __slots__ = ("_advertised",)

    def __init__(self) -> None:
        self._advertised: Dict[int, Set[str]] = {}

    def advertise(self, member_id: int, capabilities: Iterable[str]) -> None:
        """Record an advertisement (replaces any prior one)."""
        self._advertised[member_id] = set(capabilities)

    def forget(self, member_id: int) -> None:
        """Remove a member's advertisement."""
        self._advertised.pop(member_id, None)

    def capabilities_of(self, member_id: int) -> Optional[Set[str]]:
        """Current advertisement for ``member_id``."""
        s = self._advertised.get(member_id)
        return set(s) if s is not None else None

    def group_supports(self, capability: str) -> bool:
        """True iff every advertised member supports ``capability``."""
        if not self._advertised:
            return False
        return all(capability in s for s in self._advertised.values())

    def intersection(self) -> Set[str]:
        """Capabilities every member advertises (safe-to-use set)."""
        if not self._advertised:
            return set()
        sets = iter(self._advertised.values())
        acc = set(next(sets))
        for s in sets:
            acc &= s
        return acc

    def union(self) -> Set[str]:
        """Every capability advertised by any member."""
        acc: Set[str] = set()
        for s in self._advertised.values():
            acc |= s
        return acc

    def missing(self, capability: str) -> List[int]:
        """Members that did not advertise ``capability``."""
        return [m for m, s in self._advertised.items() if capability not in s]

    def __len__(self) -> int:
        return len(self._advertised)
